cvmfs/catalog_mgr_client.cc

```diff
@@ -0,0 +1,262 @@
+/**
+ * This file is part of the CernVM file system.
+ */
+
+#include "cvmfs_config.h"
+#include "catalog_mgr_client.h"
+
+#include "cache.h"
+#include "download.h"
+#include "fetch.h"
+#include "manifest.h"
+#include "quota.h"
+#include "signature.h"
+#include "statistics.h"
+
+using namespace std;  // NOLINT
+
+namespace catalog {
+
+/**
+ * Triggered when the catalog is attached (db file opened)
+ */
+void ClientCatalogManager::ActivateCatalog(Catalog *catalog) {
+  const Counters &counters = const_cast<const Catalog*>(catalog)->GetCounters();
+  if (catalog->IsRoot()) {
+    all_inodes_ = counters.GetAllEntries();
+  }
+  loaded_inodes_ += counters.GetSelfEntries();
+}
+
+
+ClientCatalogManager::ClientCatalogManager(
+  const string &repo_name,
+  cvmfs::Fetcher *fetcher,
+  signature::SignatureManager *signature_mgr,
+  perf::Statistics *statistics)
+  : AbstractCatalogManager(statistics)
+  , repo_name_(repo_name)
+  , fetcher_(fetcher)
+  , signature_mgr_(signature_mgr)
+  , offline_mode_(false)
+  , all_inodes_(0)
+  , loaded_inodes_(0)
+{
+  LogCvmfs(kLogCatalog, kLogDebug, "constructing client catalog manager");
+  n_certificate_hits_ = statistics->Register("cache.n_certificate_hits",
+    "Number of certificate hits");
+  n_certificate_misses_ = statistics->Register("cache.n_certificate_misses",
+    "Number of certificate misses");
+}
+
+
+ClientCatalogManager::~ClientCatalogManager() {
+  LogCvmfs(kLogCache, kLogDebug, "unpinning / unloading all catalogs");
+
+  for (map<PathString, shash::Any>::iterator i = mounted_catalogs_.begin(),
+       iend = mounted_catalogs_.end(); i != iend; ++i)
+  {
+    fetcher_->cache_mgr()->quota_mgr()->Unpin(i->second);
+  }
+}
+
+
+Catalog *ClientCatalogManager::CreateCatalog(
+  const PathString  &mountpoint,
+  const shash::Any  &catalog_hash,
+  catalog::Catalog  *parent_catalog
+) {
+  mounted_catalogs_[mountpoint] = loaded_catalogs_[mountpoint];
+  loaded_catalogs_.erase(mountpoint);
+  return new Catalog(mountpoint, catalog_hash, parent_catalog);
+}
+
+
+shash::Any ClientCatalogManager::GetRootHash() {
+  ReadLock();
+  shash::Any result = mounted_catalogs_[PathString("", 0)];
+  Unlock();
+  return result;
+}
+
+
+/**
+ * Specialized initialization that uses a fixed root hash.
+ */
+bool ClientCatalogManager::InitFixed(const shash::Any &root_hash) {
+  LogCvmfs(kLogCatalog, kLogDebug, "Initialize catalog with root hash %s",
+           root_hash.ToString().c_str());
+  WriteLock();
+  bool attached = MountCatalog(PathString("", 0), root_hash, NULL);
+  Unlock();
+
+  if (!attached) {
+    LogCvmfs(kLogCatalog, kLogDebug, "failed to initialize root catalog");
+  }
+
+  return attached;
+}
+
+
+LoadError ClientCatalogManager::LoadCatalog(
```

Comment:
By properly wrapping this into the `Fetcher` object. `CatalogManager` would request a `Catalog` (note: it should obtain a `Catalog` object from `Fetcher` not a file descriptor or a file path!) which it could use right away. `Fetcher` needs to do the right things for retrieving and setting up a catalog like *pin* it in the cache.

The `Catalog` itself should be initialised with a file object (maybe the `Sink`?) pointing to the SQLite file to be used. This way also the interface between `Fetcher` and `Catalog` is minimal and could even be abstracted by templating the `Fetcher` with `CatalogT`. Hence, one could easily reuse the `Fetcher` for `WritableCatalogs` in the server, for example.

When the `CatalogManager` needs to detach a loaded `Catalog`, it would remove it from its internal structures and just *delete* it. Since `Catalog` has its SQLite file wrapped by some smart object (somewhat similar to the `Sink` introduced further down) it can delegate the cache cleanup (*unpin* and friends) to a central place. Note that this way also the `Catalog` doesn't need to have the information of how to deal with its underlying SQLite file.

This is a bit out of the blue of course, but it would potentially decouple the `CatalogManager` completely from the fetching logic (and also the internal `Catalog` logic by the way). I don't see why this would require a rewrite of CVMFS.